tests/unit/utilities/package-cache-test.js

```diff
@@ -0,0 +1,384 @@
+    expect(function() { testPackageCache._writeManifest('yarn', 'yarn', manifest); }).to.not.throw(Error);
+
+    // Add a "lockfile".
+    fs.writeFileSync(lockFileLocation, 'Hello, world!');
+    expect(file(lockFileLocation)).to.exist; // Sanity check.
+
+    // Make sure it gets removed.
+    testPackageCache._writeManifest('yarn', 'yarn', manifest);
+    expect(file(lockFileLocation)).to.not.exist;
+
+    testPackageCache.destroy('bower');
+    testPackageCache.destroy('yarn');
+  });
+
+  it('_checkManifest', function() {
+    var manifest = JSON.stringify({
+      "name": "foo",
+      "dependencies": {
+        "ember": "2.9.0",
+        "ember-cli-shims": "0.1.3"
+      }
+    });
+
+    var manifestShuffled = JSON.stringify({
+      "name": "foo",
+      "dependencies": {
+        "ember-cli-shims": "0.1.3",
+        "ember": "2.9.0"
+      }
+    });
+
+    testPackageCache._writeManifest('bower', 'bower', manifest);
+
+    expect(testPackageCache._checkManifest('bower', 'bower', manifest)).to.be.true;
+    expect(testPackageCache._checkManifest('bower', 'bower', manifestShuffled)).to.be.true;
+    expect(testPackageCache._checkManifest('bower', 'bower', '{ "dependencies": "different" }')).to.be.false;
+
+    testPackageCache.destroy('bower');
+  });
+
+  it('_install', function() {
+    // We're only going to test the invocation pattern boundary.
+    // Don't want to wait for the install to execute.
+
+    // Fake in the dir label.
+    testPackageCache._conf.set('label', 'hello');
+
+    // Trigger install.
+    invocations = [];
+    testPackageCache._install('label', 'npm');
+    expect(invocations.length).to.equal(1);
+    expect(invocations[0][1]).to.equal('install');
+    expect(invocations[0][2]).to.deep.equal({ cwd: 'hello' });
+
+    // We want to make sure it attempts to link when it is supposed to.
+    invocations = [];
+    testPackageCache.options.linkEmberCLI = true;
+    testPackageCache._install('label', 'npm');
+    expect(invocations.length).to.equal(2);
+    expect(invocations[0][1]).to.equal('install');
+    expect(invocations[0][2]).to.deep.equal({ cwd: 'hello' });
+    expect(invocations[1][1]).to.equal('link');
+    expect(invocations[1][2]).to.equal('ember-cli');
+    expect(invocations[1][3]).to.deep.equal({ cwd: 'hello' });
+
+    // Make sure `bower` doesn't trigger link.
+    invocations = [];
+    testPackageCache.options.linkEmberCLI = true;
+    testPackageCache._install('label', 'bower');
+    expect(invocations.length).to.equal(1);
+    expect(invocations[0][1]).to.equal('install');
+    expect(invocations[0][2]).to.deep.equal({ cwd: 'hello' });
+  });
+
+  describe('_upgrade', function() {
+    // We're only going to test the invocation pattern boundary.
+    // Don't want to wait for the install to execute.
+    var testCounter = 0;
+    var label;
+
+    // This accounts for the downgrade ability.
+    var command;
+    if (process.version.indexOf('v0.12') === 0) {
```

Comment:
This doesn't play nicely in a loop because each package manager is slightly different. It ends up becoming brittle. Split into three separate tests.